wakuv2/nwaku.go

```diff
@@ -2451,117 +2416,54 @@ func newWakuNode(ctx context.Context, config *WakuConfig) (*WakuNode, error) {
-			numPeers, err := n.peerExchangeRequest(req.input.(uint64))
-			req.responseCh <- response{value: numPeers, err: err}
-		case requestTypeRelaySubscribe:
-			req.responseCh <- response{err: n.relaySubscribe(req.input.(string))}
-		case requestTypeRelayUnsubscribe:
-			req.responseCh <- response{err: n.relayUnsubscribe(req.input.(string))}
-		case requestTypeConnect:
-			req.responseCh <- response{err: n.connect(req.input.(connectRequest))}
-		case requestTypeDialPeerByID:
-			req.responseCh <- response{err: n.dialPeerById(req.input.(dialPeerByIDRequest))}
-		case requestTypeListenAddresses:
-			addrs, err := n.listenAddresses()
-			req.responseCh <- response{value: addrs, err: err}
-		case requestTypeENR:
-			enr, err := n.enr()
-			req.responseCh <- response{value: enr, err: err}
-		case requestTypeListPeersInMesh:
-			numPeers, err := n.listPeersInMesh(req.input.(string))
-			req.responseCh <- response{value: numPeers, err: err}
-		case requestTypeGetConnectedPeers:
-			peers, err := n.getConnectedPeers()
-			req.responseCh <- response{value: peers, err: err}
-		case requestTypeGetPeerIDsFromPeerStore:
-			peers, err := n.getPeerIDsFromPeerStore()
-			req.responseCh <- response{value: peers, err: err}
-		case requestTypeGetPeerIDsByProtocol:
-			peers, err := n.getPeerIDsByProtocol(req.input.(libp2pproto.ID))
-			req.responseCh <- response{value: peers, err: err}
-		case requestTypeDisconnectPeerByID:
-			req.responseCh <- response{err: n.disconnectPeerByID(req.input.(peer.ID))}
-		case requestTypeDnsDiscovery:
-			addrs, err := n.dnsDiscovery(req.input.(dnsDiscoveryRequest))
-			req.responseCh <- response{value: addrs, err: err}
-		case requestTypeDialPeer:
-			req.responseCh <- response{err: n.dialPeer(req.input.(dialPeerRequest))}
-		case requestTypeGetNumConnectedRelayPeers:
-			numPeers, err := n.getNumConnectedRelayPeers(req.input.([]string)...)
-			req.responseCh <- response{value: numPeers, err: err}
-		default:
-			req.responseCh <- response{err: errors.New("invalid operation")}
-		}
+	var cJsonConfig = C.CString(string(jsonConfig))
+
+	var resp = C.allocResp(unsafe.Pointer(&wg))
+
+	defer C.free(unsafe.Pointer(cJsonConfig))
+	defer C.freeResp(resp)
+
+	if C.getRet(resp) != C.RET_OK {
+		errMsg := "error wakuNew: " + C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
+		return nil, errors.New(errMsg)
 	}
+
+	wg.Add(1)
+	n.wakuCtx = C.cGoWakuNew(cJsonConfig, resp)
+	wg.Wait()
+
+	// Notice that the events for self node are handled by the 'MyEventCallback' method
+	C.cGoWakuSetEventCallback(n.wakuCtx)
+
+	return n, nil
 }
 
-func (n *WakuNode) getNumConnectedRelayPeers(optPubsubTopic ...string) (int, error) {
+func (n *WakuNode) GetNumConnectedRelayPeers(optPubsubTopic ...string) (int, error) {
 	var pubsubTopic string
 	if len(optPubsubTopic) == 0 {
 		pubsubTopic = ""
 	} else {
 		pubsubTopic = optPubsubTopic[0]
 	}
 
-	var resp = C.allocResp()
-	var cPubsubTopic = C.CString(pubsubTopic)
+	wg := sync.WaitGroup{}
```

Comment:
libwaku executes the callbacks asynchronously. This ends up causing some unintended behavior because you would expect that a call like this, to return a value:
```
WakuListPeers(....) int
```
However, the way libwaku works, it is not possible to get a value immediatly, since the waku thread requests are processed async. So, in order to get the behaviour I describe, I propose using a waitgroup. What this waitgroup will do is indicate that there's an active task that must be "waited" before continuing the execution of the code (hence why I have a wg.Wait() inside the functions), then, the Go callback will call `Done` on the waitgroup (whose reference I pass via the `Resp` struct) to indicate that the callback was executed.  This will cause the `Wait()` to proceed normally, achieving that way sync behavior for async functions!